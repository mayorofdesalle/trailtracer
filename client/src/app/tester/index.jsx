import '@testing-library/jest-dom';

import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

import './mocks/i18n.mock';
import './mocks/matchMedia.mock';
import './mocks/resizeObserver.mock';

import store from '@app/store';
import GlobalStyle from '@components/style/GlobalStyle';
import Theme from '@components/style/Theme';

const Providers = ({ children }) => {
    return (
        <Provider store={store}>
            <Theme>
                <GlobalStyle />
                <BrowserRouter>
                    {children}
                </BrowserRouter>
            </Theme>
        </Provider>
    );
};

const customRender = (ui, options) =>
    render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
export * from 'vitest';

// override render method
export { customRender as render };
