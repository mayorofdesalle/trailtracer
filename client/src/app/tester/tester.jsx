import { render } from '@testing-library/react'
import { Provider } from 'react-redux';

import './mocks/matchMedia.mock';
import store from '@app/store';
import Theme from '@components/style/Theme'

const Providers = ({ children }) => {
    return (
        <Provider store={store}>
            <Theme>
                {children}
            </Theme>
        </Provider>
    );
};

const customRender = (ui, options) =>
    render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
export * from 'vitest';

// override render method
export { customRender as render };