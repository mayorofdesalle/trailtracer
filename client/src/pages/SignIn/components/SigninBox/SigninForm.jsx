import styled, { useTheme } from 'styled-components';

import Input from '@components/ui/Input';
import Button from '@components/ui/Button';
import Text from '@components/ui/Text';
import Container from '@components/ui/Container';

const FormInner = styled.form`
    display: flex;
    flex-direction: column;
    justify-content: space-around;
    align-items: center;
    height: fit-content;
    
    & > * {
        margin-top: clamp(1.5625rem, min(3dvw, 3dvh), 3.25rem);
    }

    & > ${Container}:first-of-type {
        margin-top: 0;
    }

    & > ${Button} {
        padding: 0.25rem;
        width: 100%;
    }
`;

const SigninForm = () => {
    const theme = useTheme();

    return (
        <FormInner>
            <>
                <label hidden>Username or email</label>
                <Input type='email' name='email' autoComplete='email' placeholder='Username or email' icon='at-fill' iconColor={theme.colors.primary} required />
            </>
            <>
                <label hidden>Password</label>
                <Input type='password' eyeColor={theme.colors.secondary} name='password' autoComplete='current-password' placeholder='Password' icon='key-fill' iconColor={theme.colors.primary} required />
            </>
            <Button $bgColor={theme.colors.background} $animColors={[theme.colors.background, theme.colors.secondary]}>
                <Text $color={theme.colors.primary} $heading>SIGN IN</Text>
            </Button>
        </FormInner>
    );
};

export default SigninForm;