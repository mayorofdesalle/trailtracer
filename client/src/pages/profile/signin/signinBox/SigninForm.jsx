import { useMemo, useState, useCallback } from 'react';
import { useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import styled, { useTheme } from 'styled-components';

import { logger } from '@app';
import { Button, Container, Icon, Text } from '@components/ui';
import { PasswordInput, TextInput, changeFocus } from '@components/ui/input';
import { FormContext } from '@context';

const SigninFormInner = styled.form`
    position: relative;
    display: flex;
    flex-direction: column;
    justify-content: space-around;
    align-items: center;
    
    & > * {
        margin-top: clamp(1rem, min(3dvw, 3dvh), 3rem);
    }

    & > ${Container}:first-of-type {
        margin-top: 0;
    }

    & > ${Button} {
        padding: 0.25rem;
        width: 100%;
    }
`;

/**
 * SigninForm
 * @description
 * Form to sign in with email and password.
 * **/
const SigninForm = (props) => {
    const theme = useTheme();
    const { t } = useTranslation();

    const { register, handleSubmit, setFocus, setValue, watch, formState: { errors } } = useForm({
        defaultValues: useMemo(() => {
            return {
                email: '',
                'current-password': ''
            };
        }, []),
        mode: 'onBlur',
        reValidateMode: 'onBlur',
        shouldFocusError: true
    });

    const [isSubmitting, setIsSubmitting] = useState(false);

    const onSubmit = useCallback((data) => {
        setIsSubmitting(true);
        logger.debug(data);
    });

    return (
        <FormContext.Provider value={{ register, setFocus, setValue, watch }}>
            <SigninFormInner role='form' onSubmit={handleSubmit(onSubmit)} onKeyDown={(e) => changeFocus(e, ['email', 'current-password'], setFocus)} {...props}>
                <TextInput type='email' error={errors.email} required validate>
                    <Icon name='at-fill' color={errors.email ? theme.colors.secondary : theme.colors.primary} />
                </TextInput>
                <PasswordInput error={errors.password} required>
                    <Icon name='key-fill' color={errors.password ? theme.colors.secondary : theme.colors.primary} />
                </PasswordInput>
                <Button $type='submit' $bgColor={theme.colors.background} $animColors={[theme.colors.background, theme.colors.secondary]} disabled={isSubmitting}>
                    <Text $color={theme.colors.primary} $heading>{t('signinPage.signin')}</Text>
                </Button>
            </SigninFormInner>
        </FormContext.Provider>
    );
};

export default SigninForm;