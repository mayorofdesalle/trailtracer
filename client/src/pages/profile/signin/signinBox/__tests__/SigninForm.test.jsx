import { describe, expect, it, render, screen } from '@app/tester';

import SigninForm from '../SigninForm';

describe('SigninForm', () => {

    it('should render correctly', () => {
        render(<SigninForm />);

        const form = screen.getByRole('form');
        expect(form).toBeInTheDocument();

        const email = screen.getByRole('textbox');
        expect(email).toBeInTheDocument();

        const password = screen.getByLabelText(/password/i);
        expect(password).toBeInTheDocument();

        const button = screen.getByRole('button');
        expect(button).toBeInTheDocument();
    });

    it('should match snapshot', () => {
        const { asFragment } = render(<SigninForm />);
        expect(asFragment()).toMatchSnapshot();
    });

});