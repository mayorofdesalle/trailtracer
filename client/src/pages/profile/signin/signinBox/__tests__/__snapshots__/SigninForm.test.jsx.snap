// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SigninForm > should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="sc-eTNRI gkGkaI"
    role="form"
  >
    <div
      class="sc-guDLey sc-irLvIq kHwwsn bHzjpc"
    >
      <label
        for="email"
        hidden=""
      >
        forms.email
      </label>
      <input
        autocomplete="email"
        class="sc-kFCroH kdoZoe"
        data-tooltip-id="email"
        id="email"
        name="email"
        placeholder="forms.email"
        type="email"
      />
      <svg
        class="sc-eDLKkx fuXsHX"
        fill="#435049"
        role="img"
      >
        <title>
          at-fill
        </title>
        <use
          href="http://localhost:3000/src/assets/icons.svg#at-fill"
        />
      </svg>
    </div>
    <div
      class="sc-guDLey sc-irLvIq kHwwsn bHzjpc"
    >
      <label
        for="password"
        hidden=""
      >
        forms.current-password
      </label>
      <input
        autocomplete="current-password"
        class="sc-kFCroH gCiziP"
        data-tooltip-id="password"
        id="password"
        name="current-password"
        placeholder="forms.current-password"
        type="password"
      />
      <svg
        class="sc-eDLKkx fuXsHX"
        fill="#435049"
        role="img"
      >
        <title>
          key-fill
        </title>
        <use
          href="http://localhost:3000/src/assets/icons.svg#key-fill"
        />
      </svg>
    </div>
    <button
      class="sc-beySPh jLMfrb"
      type="submit"
    >
      <div
        class="sc-gLLuof kEDizj"
      >
        signinPage.signin
      </div>
    </button>
  </form>
</DocumentFragment>
`;
